From be9df940e0af58766f74043fafb7b4f5f3289ae9 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:33:52 -0400
Subject: [alsa] hda: Pop noises fix for XPS13 9333

Message-id: <140972823573-317-patchset-jkysela@redhat.com>
Patchwork-id: 91013
O-Subject: [RHEL 7.1 PATCH 316/442] ALSA: hda - Pop noises fix for XPS13 9333
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

When headphones are plugged in, force AFG and node 0x02
("Headphone Playback Volume") to D0 to avoid pop noises.

Bugzilla: https://bugzilla.kernel.org/show_bug.cgi?id=76611
Signed-off-by: Gabriele Mazzotta <gabriele.mzt@gmail.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 033b0a7ca9ca6589003e7cd135d2de490d3848ad)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/patch_realtek.c b/sound/pci/hda/patch_realtek.c
index 992949c..08bb069 100644
--- a/sound/pci/hda/patch_realtek.c
+++ b/sound/pci/hda/patch_realtek.c
@@ -3986,6 +3986,29 @@ static void alc_fixup_disable_aamix(struct hda_codec *codec,
  }
 }
 
+static unsigned int alc_power_filter_xps13(struct hda_codec *codec,
+    hda_nid_t nid,
+    unsigned int power_state)
+{
+ struct alc_spec *spec = codec->spec;
+
+ /* Avoid pop noises when headphones are plugged in */
+ if (spec->gen.hp_jack_present)
+  if (nid == codec->afg || nid == 0x02)
+   return AC_PWRST_D0;
+ return power_state;
+}
+
+static void alc_fixup_dell_xps13(struct hda_codec *codec,
+    const struct hda_fixup *fix, int action)
+{
+ if (action == HDA_FIXUP_ACT_PROBE) {
+  struct alc_spec *spec = codec->spec;
+  spec->shutup = alc_no_shutup;
+  codec->power_filter = alc_power_filter_xps13;
+ }
+}
+
 static void alc_fixup_headset_mode_alc668(struct hda_codec *codec,
     const struct hda_fixup *fix, int action)
 {
@@ -5485,6 +5508,7 @@ enum {
  ALC662_FIXUP_BASS_CHMAP,
  ALC668_FIXUP_AUTO_MUTE,
  ALC668_FIXUP_DELL_DISABLE_AAMIX,
+ ALC668_FIXUP_DELL_XPS13,
 };
 
 static const struct hda_fixup alc662_fixups[] = {
@@ -5651,6 +5675,12 @@ static const struct hda_fixup alc662_fixups[] = {
   .type = HDA_FIXUP_FUNC,
   .v.func = alc_fixup_inv_dmic_0x12,
  },
+ [ALC668_FIXUP_DELL_XPS13] = {
+  .type = HDA_FIXUP_FUNC,
+  .v.func = alc_fixup_dell_xps13,
+  .chained = true,
+  .chain_id = ALC668_FIXUP_DELL_DISABLE_AAMIX
+ },
  [ALC668_FIXUP_DELL_DISABLE_AAMIX] = {
   .type = HDA_FIXUP_FUNC,
   .v.func = alc_fixup_disable_aamix,
@@ -5717,7 +5747,7 @@ static const struct snd_pci_quirk alc662_fixup_tbl[] = {
  SND_PCI_QUIRK(0x1025, 0x038b, "Acer Aspire 8943G", ALC662_FIXUP_ASPIRE),
  SND_PCI_QUIRK(0x1028, 0x05d8, "Dell", ALC668_FIXUP_DELL_MIC_NO_PRESENCE),
  SND_PCI_QUIRK(0x1028, 0x05db, "Dell", ALC668_FIXUP_DELL_MIC_NO_PRESENCE),
- SND_PCI_QUIRK(0x1028, 0x060a, "Dell XPS 13", ALC668_FIXUP_DELL_DISABLE_AAMIX),
+ SND_PCI_QUIRK(0x1028, 0x060a, "Dell XPS 13", ALC668_FIXUP_DELL_XPS13),
  SND_PCI_QUIRK(0x1028, 0x0623, "Dell", ALC668_FIXUP_AUTO_MUTE),
  SND_PCI_QUIRK(0x1028, 0x0624, "Dell", ALC668_FIXUP_AUTO_MUTE),
  SND_PCI_QUIRK(0x1028, 0x0625, "Dell", ALC668_FIXUP_DELL_MIC_NO_PRESENCE),
-- 
1.7.1